src/electrum.rs

```diff
@@ -275,6 +275,71 @@ impl Rpc {
+            Ok(tx) => tx,
+            Err(_) => return Ok(json!({})),
+        };
+
+        let funding_inputs = &funding_tx.input;
+
+        if funding_blockhash.is_none() && funding_inputs.iter().any(|txi| self.tracker.get_blockhash_by_txid(txi.previous_output.txid).is_none()) {
+            funding_inputs_confirmed = false;
+        }
+
+        let funding_height = match &funding_blockhash {
+            Some(funding_blockhash) => self.tracker.chain().get_block_height(funding_blockhash).ok_or_else(|| anyhow::anyhow!("Blockhash not found"))?,
+            None => 0,
+        };
+
+        let tx_candidates: Vec<Txid> = match spending_blockhash {
+            None => self.tracker.mempool().filter_by_spending(&funding).iter().map(|e| e.txid).collect(),
+            Some(spending_blockhash) => {
+                let mut txids: Vec<Txid> = Vec::new();
+                self.daemon.for_blocks(Some(spending_blockhash).into_iter(), |_, block| {
+                    for tx in block.txdata.into_iter() {
+                        if is_spending(&tx, funding) {
+                            txids.push(tx.txid())
+                        }
+                    }
+                })?;
+                txids
+            },
+        };
+        
+        let mut spender_txids = tx_candidates.iter();
+
+        match spender_txids.next() {
+            Some(spender_txid) => if let Some(double_spending_txid) = spender_txids.next() {
+                    return bail!("double spend of {}: {}", spender_txid, double_spending_txid)
+                } else {
+                    match spending_blockhash {
+                        Some(spending_blockhash) => {
+                            let spending_height = self.tracker.chain().get_block_height(&spending_blockhash).ok_or_else(|| anyhow::anyhow!("Blockhash not found"))?;
+                            return Ok(json!({"height": funding_height, "spender_txhash": spender_txid, "spender_height": spending_height}))
+                        }
+                        None => {
+                            let spending_tx = self.daemon.get_transaction(&spender_txid, None)?;
+                            if spending_tx.input.iter().any(|txi| self.tracker.get_blockhash_by_txid(txi.previous_output.txid).is_none()) {
+                                if funding_inputs_confirmed {
+                                    return Ok(json!({"height": funding_height, "spender_txhash": spender_txid, "spender_height": -1}));
+                                }
+                                return Ok(json!({"height": -1, "spender_txhash": spender_txid, "spender_height": -1}));
+                            } else {return Ok(json!({"height": funding_height, "spender_txhash": spender_txid, "spender_height": 0}));}
+                        }
+                }
+            },
+            None => return Ok(json!({"height": funding_height})),
+        };
```

Comment:
I think this whole match could be cleaned up a bit:

```rust
let spender_txid = spender_txids.next();
let double_spending_txid  = spender_txids.next(); // slice-based operator is fused, so this is OK
match (spender_txid, double_spending_txid, spending_blockhash) {
    (Some(spender_txid), Some(double_spending_txid), _) => bail!("double spend of {}: {}", spender_txid, double_spending_txid),
    (Some(spender_txid), None, Some(spending_blockhash)) => {
        let spending_height = self.tracker.chain().get_block_height(&spending_blockhash).ok_or_else(|| anyhow::anyhow!("Blockhash not found"))?;
        Ok(json!({"height": funding_height, "spender_txhash": spender_txid, "spender_height": spending_height}))
    },
    (Some(spender_txid), None, None) => {
let spending_tx = self.daemon.get_transaction(&spender_txid, None)?;
        if funding_inputs_confirmed {
            if spending_tx.input.iter().any(|txi| self.tracker.get_blockhash_by_txid(txi.previous_output.txid).is_none()) {
                Ok(json!({"height": funding_height, "spender_txhash": spender_txid, "spender_height": -1}))
            } else {
                Ok(json!({"height": funding_height, "spender_txhash": spender_txid, "spender_height": 0}))
            }
        } else {
            Ok(json!({"height": -1, "spender_txhash": spender_txid, "spender_height": -1}))
        }
    },
}
```

I might have left some syntax errors. Not that aside from using single match and removing returns I also swapped the conditions so that `spending_tx.input.iter().any(|txi| self.tracker.get_blockhash_by_txid(txi.previous_output.txid).is_none())` is **not** computed when `funding_inputs_confirmed` is false.
If funding inputs are unconfirmed then funding transaction is definitely also unconfirmed and then spender is unconfirmed too so we don't need to calculate it in that case. Swapping them as I did is thus more efficient.

Reply:
Oow I didn't realise how powerful match can be :O this is indeed much better ! And it is much easier to handle the case where spending transaction is not found in the block the spending index return. Cool stuff, I will include it.

In fact I can even replace `if funding_inputs_confirmed` by `if !(funding_blockhash.is_none() && funding_inputs.iter().any(|txi| self.tracker.get_blockhash_by_txid(txi.previous_output.txid).is_none()))` directly since it isn't used anywhere else.

Then the gain efficiency of the swap comes from the fact that if funding is not confirmed, this test will not look at its inputs to return false and we don't even need to look at spending transaction inputs too ! This is not possible when testing spending inputs first indeed.

Reply:
Done !

I had to keep funding_inputs_confirmed in the end because we also use it if the spending transaction doesn't exist at all which is another match case.

It also handle the case of bad indexing.